File: PE-diagram/src/main/java/nl/adaptivity/process/clientProcessModel/ClientStartNode.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.clientProcessModel;

import nl.adaptivity.process.processModel.*;
import nl.adaptivity.process.util.Identifiable;
import nl.adaptivity.process.util.Identified;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;


public class ClientStartNode<NodeT extends ClientProcessNode<NodeT, ModelT>, ModelT extends ClientProcessModel<NodeT, ModelT>> extends StartNodeBase<NodeT, ModelT> implements ClientProcessNode<NodeT, ModelT> {

  public static class Builder<NodeT extends ClientProcessNode<NodeT, ModelT>, ModelT extends ClientProcessModel<NodeT, ModelT>> extends StartNodeBase.Builder<NodeT, ModelT> implements ClientProcessNode.Builder<NodeT, ModelT> {

    public Builder() { }

    public Builder(final boolean compat) {
      this.compat = compat;
    }

    public Builder(@Nullable final Identified successor, @Nullable final String id, @Nullable final String label, final double x, final double y, @NotNull final Collection<? extends IXmlDefineType> defines, @NotNull final Collection<? extends IXmlResultType> results) {
      super(id, successor, label, defines, results, x, y);
    }

    public Builder(@NotNull final StartNode<?, ?> node) {
      super(node);
      if (node instanceof ClientStartNode) {
        compat = ((ClientStartNode) node).isCompat();
      } else {
        compat = false;
      }
    }

    @NotNull
    @Override
    public ClientStartNode<NodeT, ModelT> build(@NotNull final ModelT newOwner) {
      return new ClientStartNode<NodeT, ModelT>(this, newOwner);
    }

    @Override
    public boolean isCompat() {
      return compat;
    }

    @Override
    public void setCompat(final boolean compat) {
      this.compat = compat;
    }

    public boolean compat = false;
  }

  private final boolean mCompat;

  public ClientStartNode(final ModelT ownerModel, final boolean compat) {
    super(ownerModel);
    mCompat = compat;
  }

  public ClientStartNode(final ModelT ownerModel, final String id, final boolean compat) {
    super(ownerModel);
    setId(id);
    mCompat = compat;
  }

  protected ClientStartNode(final ClientStartNode<NodeT, ModelT> orig, final boolean compat) {
    super(orig.builder(), null);
    mCompat = compat;
  }

  public ClientStartNode(@NotNull final StartNode.Builder<?, ?> builder, @NotNull final ModelT newOwnerModel) {
    super(builder, newOwnerModel);
    if (builder instanceof Builder) {
      mCompat = ((Builder) builder).compat;
    } else {
      mCompat = false;
    }
  }

  @NotNull
  @Override
  public Builder<NodeT, ModelT> builder() {
    return new Builder<>(this);
  }

  @Override
  public void setId(@Nullable final String id) {
    super.setId(id);
  }

  @Override
  public int getMaxSuccessorCount() {
    return isCompat() ? Integer.MAX_VALUE : 1;
  }

  @Override
  public boolean isCompat() {
    return mCompat;
  }

  @Override
  public void setOwnerModel(@NotNull final ModelT newOwnerModel) {
    super.setOwnerModel(newOwnerModel);
  }

  @Override
  public void setPredecessors(@NotNull final Collection<? extends Identifiable> predecessors) {
    super.setPredecessors(predecessors);
  }

  @Override
  public void removePredecessor(@NotNull final Identified node) {
    super.removePredecessor(node);
  }

  @Override
  public void addPredecessor(final Identified nodeId) {
    super.addPredecessor(nodeId);
  }

  @Override
  public void addSuccessor(final Identified node) {
    super.addSuccessor(node);
  }

  @Override
  public void removeSuccessor(@NotNull final Identified node) {
    super.removeSuccessor(node);
  }

  @Override
  public void setSuccessors(@NotNull final Collection<? extends Identified> successors) {
    super.setSuccessors(successors);
  }

}

File: PE-diagram/src/main/kotlin/nl.adaptivity.process/clientProcessModel/ClientStartNode.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.clientProcessModel

import nl.adaptivity.process.processModel.IXmlDefineType
import nl.adaptivity.process.processModel.IXmlResultType
import nl.adaptivity.process.processModel.StartNode
import nl.adaptivity.process.processModel.StartNodeBase
import nl.adaptivity.process.util.Identifiable
import nl.adaptivity.process.util.Identified

open class ClientStartNode : StartNodeBase<NodeT, ModelT>, ClientProcessNode {

    open class Builder : StartNodeBase.Builder<NodeT, ModelT>, ClientProcessNode.Builder {

        constructor() {}

        constructor(compat: Boolean) {
            this.isCompat = compat
        }

        constructor(successor: Identified?, id: String?, label: String?, x: Double, y: Double, defines: Collection<IXmlDefineType>, results: Collection<IXmlResultType>) : super(id, successor, label, defines, results, x, y) {}

        constructor(node: StartNode<*, *>) : super(node) {
            if (node is ClientStartNode) {
                isCompat = node.isCompat
            } else {
                isCompat = false
            }
        }

        override fun build(newOwner: ModelT): ClientStartNode {
            return ClientStartNode(this, newOwner)
        }

        override var isCompat = false
    }

    override val isCompat: Boolean

    constructor(ownerModel: ModelT, compat: Boolean) : super(ownerModel) {
        isCompat = compat
    }

    constructor(ownerModel: ModelT, id: String, compat: Boolean) : super(ownerModel, id=id) {
        isCompat = compat
    }

    protected constructor(orig: ClientStartNode, newOwnerModel: ModelT, compat: Boolean) : super(orig.builder(), newOwnerModel) {
        isCompat = compat
    }

    constructor(builder: StartNode.Builder<*, *>, newOwnerModel: ModelT) : super(builder, newOwnerModel) {
        if (builder is Builder) {
            isCompat = builder.isCompat
        } else {
            isCompat = false
        }
    }

    override fun builder(): Builder {
        return Builder(this)
    }

    override val maxSuccessorCount: Int
        get() = if (isCompat) Integer.MAX_VALUE else 1

    override fun setId(id: String) = super.setId(id!!)

    override fun setLabel(label: String?) = super.setLabel(label)

    override fun setOwnerModel(newOwnerModel: ModelT) = super.setOwnerModel(newOwnerModel)

    override fun setPredecessors(predecessors: Collection<Identifiable>) = super.setPredecessors(predecessors)

    override fun removePredecessor(predecessorId: Identified) = super.removePredecessor(predecessorId)

    override fun addPredecessor(predecessorId: Identified) = super.addPredecessor(predecessorId)

    override fun addSuccessor(successorId: Identified) = super.addSuccessor(successorId)

    override fun removeSuccessor(successorId: Identified) = super.removeSuccessor(successorId)

    override fun setSuccessors(successors: Collection<Identified>) = super.setSuccessors(successors)

}
